### PR TITLE
jobs: use job_type column for RunningJobExists queries

### PR DESCRIPTION
**Pull request by dt**
Avoiding the virtual table avoids the join on infos.

Release note: none.
Epic: none.

### CODE REVIEW

File: pkg/jobs/utils.go [hunk 1]
```diff
@@ -60,7 +107,14 @@ ORDER BY created`
 			return false /* exists */, err
 		}
 
-		if payloadPredicate(payload) {
+		isTyp := false
+		for _, typ := range jobTypes {
+			if payload.Type() == typ {
+				isTyp = true
+			}
+			break
```
Comment (adityamaru) [inline pkg/jobs/utils.go:115]: should this break be inside the if condition? I assume we want to check the payload type against all `jobTypes` before bailing.

**Review by adityamaru — APPROVED** [review]

File: pkg/jobs/utils.go (hunk 1 above)
Comment (dt) [reply pkg/jobs/utils.go]: doh, yep.

**Review by rytaft — COMMENTED** [review]
Thanks for the quick fix!

> Reviewed 3 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102171#-:-NTorOew6rxQnBdbqV27:b-e69qx5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[`pkg/jobs/utils.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102171#-NTor3tc6WTXLsVXFmGU:-NTor3tc6WTXLsVXFmGV:b-ujtrlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/41a8d6a261d916ff6361957587e62c0e41077bce/pkg/jobs/utils.go#L74)):*
> ```Go
> 		return false, err
> 	}
> 	return ok && jobspb.JobID(*it.Cur()[0].(*tree.DInt)) != ignoreJobID, nil
> ```

why not put this check for `!= ignoreJobID` in the query itself? otherwise, I don't think we can guarantee that no other job also exists

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @miretskiy, and @rytaft)

___
*[`pkg/jobs/utils.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102171#-NTor3tc6WTXLsVXFmGU:-NTpgePsExrZR3SGgHpJ:bc0j5gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/41a8d6a261d916ff6361957587e62c0e41077bce/pkg/jobs/utils.go#L74)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why not put this check for `!= ignoreJobID` in the query itself? otherwise, I don't think we can guarantee that no other job also exists
</blockquote></details>

I almost did that but realized -- if my reading of the old function is correct -- that doing so isn't quite the same as what this function was/is doing/its intent: In the previous version of the function (preserved below for x-version case), there is a `break` -- not a `continue` if the row matches the passed ID. At first glance I thought this was a bug -- maybe the second row is a matching type with a different ID? -- but inspection of the createStats resumer call made me conclude it is intentional: since the query is ordered by `created`, if the "ignoredJob" is the first result, it returns false and allows it to survive/continue to execute, relying on the fact the other job, when it runs the same check, will _not_ be first and will the the one to fail, thus killing the duplicates but also ensuring _one of them_ survives. If both jobs got `true` neither would survive. Similarly if I put a filter for not the job's own ID in the query, both would see the other and fail, but checking if there are non-zero results _that start with_ a different job gets us the same behavior the old loop had?

I spent a few minutes trying to work this out and updated the comment, but see now that in a fixup I dropped my comment edit in the second commit (it had " and any jobs created after it," in the description of the ignoredID).

**Comment by dt** [comment]
Updated the comments to highlight the ordering subtly w.r.t ignored ID

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102171#-:-NTpmUq7-C6vMhQt_GO1:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)

___
*[`pkg/jobs/utils.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/102171#-NTor3tc6WTXLsVXFmGU:-NTplupI8SzGd5z_IhNo:bzamtsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/41a8d6a261d916ff6361957587e62c0e41077bce/pkg/jobs/utils.go#L74)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I almost did that but realized -- if my reading of the old function is correct -- that doing so isn't quite the same as what this function was/is doing/its intent: In the previous version of the function (preserved below for x-version case), there is a `break` -- not a `continue` if the row matches the passed ID. At first glance I thought this was a bug -- maybe the second row is a matching type with a different ID? -- but inspection of the createStats resumer call made me conclude it is intentional: since the query is ordered by `created`, if the "ignoredJob" is the first result, it returns false and allows it to survive/continue to execute, relying on the fact the other job, when it runs the same check, will _not_ be first and will the the one to fail, thus killing the duplicates but also ensuring _one of them_ survives. If both jobs got `true` neither would survive. Similarly if I put a filter for not the job's own ID in the query, both would see the other and fail, but checking if there are non-zero results _that start with_ a different job gets us the same behavior the old loop had?

I spent a few minutes trying to work this out and updated the comment, but see now that in a fixup I dropped my comment edit in the second commit (it had " and any jobs created after it," in the description of the ignoredID).
</blockquote></details>

Ah good call - thanks for the explanation!

**Comment by dt** [comment]
sadly the test cluster that prompted this change has been observe to be selecting the job_type index to run this query, instead of the status index, so it is scanning 130k rows instead of 10 for some reason, so more to do here before this is ready

**Comment by dt** [comment]
amended thanks to some help from queries: the optimizer doesn't have stats for this table, due to it being excluded in #80887, so it doesn't know that the job_type index has 137k matching entries while the job status index has 4. I've filed https://github.com/cockroachdb/cockroach/issues/102213 to track fixing the underlying issue, but for now this change has a manual hard-coded index hint since generally we expect the set of running jobs to be smaller than the set of all jobs for any given type in any cluster where either of these sets is non-trivial.

RFAL again.

**Comment by dt** [comment]
For those with access to the 23.1 testing cluster, this has been running on it with the added index hint for a bit and seems like it is a big improvement: cpu time for this query is down from 76ms to 130us, exec time from 2.6s to 23ms, and the jobs table range is no longer top of the hot ranges page for reads.

File: pkg/jobs/utils.go [hunk 2]
```diff
@@ -12,23 +12,75 @@ package jobs
...
+	jobTypes ...jobspb.Type,
+) (exists bool, retErr error) {
+	if !cv.IsActive(ctx, clusterversion.V23_1BackfillTypeColumnInJobsTable) {
+		return legacyRunningJobExists(ctx, ignoreJobID, txn, jobTypes...)
+	}
+	if len(jobTypes) == 0 {
+		return false, errors.AssertionFailedf("must specify job types")
+	}
+
+	typeStrs := fmt.Sprintf("('%s'", jobTypes[0].String())
```
Comment (knz) [inline pkg/jobs/utils.go:42]: you'd make me happier with a `strings.Builder` here.

**Review by knz — APPROVED** [review]

File: pkg/jobs/utils.go (hunk 2 above)
Comment (dt) [reply pkg/jobs/utils.go]: heh, I initially did that -- along with specializing the actually-called 1 and 2 arg cases to both be a single allocation -- but then told myself that two extra allocs here per call, once a minute, just didn't matter at least compared to actually running the query a few lines below. 

But if you had the same impulse, then that's two of us, so I put it back now.

**Comment by dt** [comment]
> TestTenantReport: test_server_shim.go:460: capabilities not yet ready

bors r+

**Comment by dt** [comment]
> acceptance/mixed-version: schema change failed.

looks like due to parse error on `VISIBILITY 0.8` index clause. unrelated.

bors r+

**Comment by knz** [comment]
> looks like due to parse error on a VISIBILITY 0.8 modifier on an index clause. unrelated.

Could you double check the issue is filed already though?

**Pull request merged**